src/analog.rs

```diff
@@ -0,0 +1,50 @@
+//! Analog-digital conversion traits
+
+use nb;
+
+/// This trait denotes an object, e.g. a GPIO pin, that is configured for use as an ADC channel.
+/// Objects of this type are passed to ADC-mode-specific methods.
+pub trait AdcChannel {
+    /// Channel ID type
+    ///
+    /// e.g. u8 for ADC channels 1-n, or a (partid, pinid) tuple for something more complex.
+    type ID;
+
+    /// ID of the channel associated with this pin
+    const CHANNEL: Self::ID;
+}
+
+/// ADCs that sample on single channels per request, and do so at the time of the request.
+pub trait SingleMode<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take an analog reading from the specified pin
+    fn read_channel(&mut self, pin: &mut Pin) -> nb::Result<Word, Self::Error>;
+}
+
+/// ADCs that sample continuously without requiring requests
+pub trait Continuous<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take ownership of an ADC channel/pin
+    fn add_channel(&mut self, pin: Pin) -> nb::Result<(), Self::Error>;
+
+    /// Release the ADC channel/pin
+    fn release_channel(&mut self) -> nb::Result<Pin, Self::Error>;
+
+    /// Start a continuous conversion cycle
+    fn start(&mut self) -> nb::Result<(), Self::Error>;
+
+    /// Stop conversions.
+    ///
+    /// If the conversions cannot be stopped immediately, returns nb::Error::WouldBlock.
+    fn stop(&mut self) -> nb::Result<(), Self::Error>;
+
+    /// Read the specified channel
+    fn read(&mut self, chan: Pin::ID) -> nb::Result<Word, Self::Error>;
```

Comment:
In addition to @austinbes's concerns, I'm confused about the overall purpose. What's the point of starting a continuous conversion, if I then have to poll the `read` method anyway?

Reply:
On such point might be to get a sample rate that's known, and regulated by hardware.

With that said, it would probably be more ergonomic (and more in-line with standard embedded paradigms) to somehow provide a buffer of samples rather than each individual one. I'm not sure the right way to do that in a memory-safe manner, though...

Reply:
The `read` method here would _likely_ be non-blocking; For example, in the (iirc) LPC178x, each channel has a specific register and the results of the most recent conversion for that specific channel are stored there. (There's also a register for last conversion result for _any_ channel.)

Reply:
Also, I contemplated a way to provide access to a buffer, but I couldn't see a way until const generics land because the number of channels would need to be constant in order to know the size of the buffer provided.

Reply:
I don't know if it's appropriate here, but I think the canonical way to work around the lack of const generics is [`generic-array`](https://crates.io/crates/generic-array).